NETOBSERV-1260 WIP lokiless blog (#55)
netobserv/documents: blogs/lokiless/lokiless.md

# Deploying Network Observability without Loki

> This blog post talks about a new feature that comes with Network Observability (NetObserv) 1.4: the ability to deploy it without Loki, coupled with exporting flows to an alternative storage. We will try to explain what this new feature means to you, users, and to us, NetObserv developers.

Reviewer: ```suggestion
This blog post talks about a new feature that comes with Network Observability (NetObserv) 1.4: the ability to deploy it without Loki, coupled with exporting flows to an alternative storage. In this article, we will desmontrate a possible configuration to consume flows outside of Loki.
```

Author: I said that because I wrote a couple of things about why we did that, what's the pros and cons, and some thoughts about the possible follow-ups that this feature enables, etc.

Reviewer: yeah the sentence can keep the "exploration" side of the article but without being too vague on the content. WDYT ?

Author: So I reformulated base on that, but trying to convey all that I want:

> It clarifies what this change is and what it isn't. Then it demonstrates the optional configuration for consuming flows outside of Loki. Finally, it explores potential avenues for future development that this feature enables.

## Why changing now?

To be clear, **we aren't actually moving away from Loki**. Loki remains the one and only storage solution that we fully support at the moment, and our Console plugin is entirely based on queries to Loki, in its _logql_ format. However, we have seen some people using NetObserv in a way that we didn't expect: deploying it without Loki, and configuring flows exporters for instance with Kafka or IPFIX. Why? It turned out they were more interested in the kube-enriched raw flow data than in the visualizations that NetObserv provides, and they didn't want to deal with a new backend storage setup and maintenance. Which, admittedly, is a quite reasonable argument.

To summarize, here's the deal:
- 💰: you save on operational aspects by not deploying Loki or any storage that would be new to you.
- 💔: you loose all the fancy dashboards that we build with so much love.

> - 👩🏽‍🔧: you need to create your own Kafka consumer for doing anything with the exported flows: this is on your plate.

Reviewer: ```suggestion
- 👩🏽‍🔧: you need to create your own consumer for doing anything with the exported flows: this is on your plate.
```

Maybe mention IPFIX else ?

Author: done (kafka+ipfix)

## What is changed

> Before 1.4, there was still a problem: NetObserv was designed with Loki as a requirement. If you don't configure the Loki endpoint, our _flowlogs-pipeline_ component still tries to send flows to a default URL (and fails), and our Console plugin still tries to query Loki (and fails). While the latter isn't too annoying for someone who anyway doesn't want to use the Console plugin, the former could be the cause of performance degradation.

Reviewer 2: This paragraph is past-tense by nature because it describes the "Before 1.4 NetObserv", so I applied consistent past-tense in this paragraph. There was some mixing of tenses. Also the same capital C console from before.

> Before 1.4, there was still a problem: NetObserv was designed with Loki as a requirement. If you don't configure the Loki endpoint, our _flowlogs-pipeline_ component still tries to send flows to a default URL (and fails), and our Console plugin still tries to query Loki (and fails). While the latter isn't too annoying for someone who anyway doesn't want to use the Console plugin, the former could be the cause of performance degradation.

Author: thanks - I'm rewritting it, with parts of your suggestion

> So this is what we did: **we "just" added an _enable_ knob for Loki**. With Loki turned off, _flowlogs-pipeline_ obviously doesn't try to send anything to it. And since the Console plugin becomes useless without Loki, it isn't deployed anymore in that case.

Reviewer 2: ```suggestion
So this is what we did: **we "just" added an _enable_ knob for Loki**. With Loki turned off, the _flowlogs-pipeline_ obviously doesn't try to send anything to it. And since the console plugin becomes useless without Loki, it isn't deployed anymore in that case.
```

Author: Saying "the _flowlogs-pipeline_" sounds weird to my ears, but I don't know if it's just me, I know this is also used in the docs. Shouldn't we consider flowlogs-pipeline as a proper noun ?

Reviewer 2: I think its OK to omit "the" if it sounds weird. I can see how it crowds the sentence. It could go either way I think..this was more of a suggestion. The is just a definite article used to increase clarity around a thing or object, and it could go before a proper noun but could also crowd the sentence if the meaning is general anyways.

![Architecture diagram](./images/arch-before-after.gif)
_NetObserv architecture diagram: before and after 1.4_

> As the diagram shows, if we choose to remove Loki, what remains of the flows pipeline downstream is:

Reviewer 2: IBM Style recommends using a complete sentence to introduce a list, rather than leaving a cliffhanger before the colon.

- The ability to generate Prometheus metrics. Those metrics and their related dashboards are still accessible in the OpenShift Console, independently from our plugin.
- The ability to set up one or several exporters downstream the pipeline, such as via Kafka or to any IPFIX collector. This is then up to you to consume this data for any purpose.

## Example use case

> Enough theory, let's put it into practice: we are going to set up NetObserv to export enriched flows to Kafka, and we will create a very simple consumer that stores them in a [ClickHouse database](https://clickhouse.com/). To do so, we created a sample application: [kafka-clickhouse-example](https://github.com/jotak/kafka-clickhouse-example/).

Reviewer 2: I guess this is a broad question, but does`NetObserv` refer to the upstream? Should we use "Network Observability" instead of "NetObserv" 🤔 ?

> Enough theory, let's put it into practice: we are going to set up NetObserv to export enriched flows to Kafka, and we will create a very simple consumer that stores them in a [ClickHouse database](https://clickhouse.com/). To do so, we created a sample application: [kafka-clickhouse-example](https://github.com/jotak/kafka-clickhouse-example/).

Author: I take the blog post format as an opportunity to run away from the corporate slang 😁
To make it clear this isn't only about the upstream, I started the post saying:

> This blog post talks about a new feature that comes with Network Observability (NetObserv) 1.4

With the intent to set the context for the rest of the article ... is that all right?
Actually, everything here works both with the upstream and the downstream versions.

> Enough theory, let's put it into practice: we are going to set up NetObserv to export enriched flows to Kafka, and we will create a very simple consumer that stores them in a [ClickHouse database](https://clickhouse.com/). To do so, we created a sample application: [kafka-clickhouse-example](https://github.com/jotak/kafka-clickhouse-example/).

Reviewer 2: Ahh, okay. Yeah I guess it is okay to call it NetObserv in the blog if we can divert from the legally approved naming in Red Hat blogs.  I guess blogs probably don't have to adhere to that. And you define it appropriately, as you said. Disregard :) I had my peer review hat on lol ..I guess in the documentation, there are legally approved acronyms for things, like some Operators can reference themselves as the acronym and some cannot.

For simplicity, ClickHouse is deployed locally and [ktunnel](https://github.com/omrikiei/ktunnel) is used for reverse port-forwarding, magically bringing ClickHouse "into" the cluster. Obviously, you won't need `ktunnel` when using a real ClickHouse server.

### Prerequisites

- An OpenShift or Kubernetes cluster
- [NetObserv operator](https://github.com/netobserv/network-observability-operator) installed (do not install a `FlowCollector` yet).
- ClickHouse binary: grab it as explained in their [quick install guide](https://clickhouse.com/docs/en/install#quick-install).
- [ktunnel](https://github.com/omrikiei/ktunnel) binary.
- Some common tools such as `curl`, `kubectl`, `envsubst`...

> Note: These steps have been tested on a small OCP 4.13 cluster with 3 worker nodes. There is no one-size-fits-all configuration, so you might need to adapt some settings depending on your environment, such as [the Kafka configuration](https://github.com/jotak/kafka-clickhouse-example/blob/main/contrib/kafka.yaml).

### Start ClickHouse with ktunnel

Using the ClickHouse binary that you downloaded, run:

```bash
./clickhouse server
```

> This will start a ClickHouse server that listens on `:9000` on your machine.

Reviewer 2: ```suggestion
This starts a ClickHouse server that listens on `:9000` on your machine.
```

# Wait that all pods are up and running, with the KafkaTopic being ready (a few minutes...)
kubectl wait --timeout=180s --for=condition=ready kafkatopic flows-export -n netobserv
kubectl get pods -n netobserv
```

When listing pods, you should see something like this:

```
NAME                                             READY   STATUS    RESTARTS   AGE
kafka-cluster-entity-operator-5f5bc5675b-lpmmc   3/3     Running   0          3m33s
kafka-cluster-kafka-0                            1/1     Running   0          4m47s
kafka-cluster-kafka-1                            1/1     Running   0          4m47s
kafka-cluster-kafka-2                            1/1     Running   0          4m47s
kafka-cluster-zookeeper-0                        1/1     Running   0          5m15s
kafka-cluster-zookeeper-1                        1/1     Running   0          5m15s
kafka-cluster-zookeeper-2                        1/1     Running   0          5m15s
strimzi-cluster-operator-6bdcd796f6-qsc4g        1/1     Running   0          17m
```

### Prepare NetObserv

> Assuming you already installed the operator, now you must create a `FlowCollector` resource that will start sending flow logs to Kafka. For the purpose of this article, we won't install and configure Loki: all we want is to get the flows into Kafka and ClickHouse.

Reviewer: ```suggestion
Assuming you already installed the operator, now you must create a `FlowCollector` resource that will start sending flow logs to Kafka. For the purpose of this article, we won't install and configure Loki but be informed that you can export flows to one or multiple endpoints and save them into Loki in parallel according to your needs.
```

> Note: we configure here Kafka as an **exporter**, which is unrelated to the `spec.deploymentModel: KAFKA` / `spec.kafka` settings: those ones correspond to NetObserv's internal flows processing configuration (NetObserv being both the producer and the consumer), whereas `spec.exporters` relates to NetObserv being just the producer, leaving up to us how we want to consume that data.

```bash
cat <<EOF | kubectl apply -f -
apiVersion: flows.netobserv.io/v1beta1
kind: FlowCollector
metadata:
  name: cluster
spec:
  namespace: netobserv
  deploymentModel: DIRECT
  loki:
    enable: false
  exporters:
    - type: KAFKA
      kafka:
        address: "kafka-cluster-kafka-bootstrap.netobserv"
        topic: flows-export
EOF
```

The Kafka `address` and `topic` configured here match what we deployed in the previous step.

> At this point, flows should be generated, collected, enriched and sent to Kafka, but with no consumer on the other end. 

Reviewer 2: You don't need a comma before this "but" because "with no consumer on the other end" is not an independent clause w/ a subject and verb. A comma is only needed before a conjunction if there's an independent clause on the other side.

### Run the consumer

> Almost all pieces are up and running, now let's bring the missing one: a Kafka consumer that will send flows to ClickHouse. This is [the sample application](https://github.com/jotak/kafka-clickhouse-example/) that we mentioned above.

Reviewer 2: ```suggestion
Almost all pieces are up and running, now let's bring the missing one: a Kafka consumer that sends flows to ClickHouse. This is [the sample application](https://github.com/jotak/kafka-clickhouse-example/) that we mentioned above.
```

To deploy it:

```bash
kubectl apply -f https://raw.githubusercontent.com/jotak/kafka-clickhouse-example/main/contrib/deployment.yaml -n netobserv
```

> Now, as everything works as expected (everything always works as expected), you will start getting the precious data in your database.

Reviewer: ```suggestion
Now, as everything works as expected (everything always works as expected), you will start getting the precious records in your database.
```

> Now, as everything works as expected (everything always works as expected), you will start getting the precious records in your database.

Reviewer 2: ```suggestion
Now, as everything works as expected (everything always works as expected), the precious records begin flowing into your database.
```

Reviewer 2: Just a suggestion to avoid future tense if you can

│ 2023-09-26 10:10:34 │ 2023-09-26 10:10:34 │ 10.129.2.14 │ 10.0.211.51 │ prometheus-k8s-1 │ ip-10-0-211-51.eu-west-3.compute.internal │ Pod      │ Node     │ openshift-monitoring │               │    66 │       1 │
└─────────────────────┴─────────────────────┴─────────────┴─────────────┴──────────────────┴───────────────────────────────────────────┴──────────┴──────────┴──────────────────────┴───────────────┴───────┴─────────┘
┌───────────────start─┬─────────────────end─┬─src_ip──────┬─dst_ip─────┬─src_name────────────────────────────┬─dst_name──────────────────────────────────┬─src_kind─┬─dst_kind─┬─src_namespace────────────────┬─dst_namespace─┬─bytes─┬─packets─┐
│ 2023-09-26 10:10:33 │ 2023-09-26 10:10:33 │ 10.129.0.13 │ 10.129.0.2 │ controller-manager-565b9fb799-vz9w9 │ ip-10-0-211-51.eu-west-3.compute.internal │ Pod      │ Node     │ openshift-controller-manager │               │    66 │       1 │
└─────────────────────┴─────────────────────┴─────────────┴────────────┴─────────────────────────────────────┴───────────────────────────────────────────┴──────────┴──────────┴──────────────────────────────┴───────────────┴───────┴─────────┘
```

> Here we go!

Reviewer: https://i.pinimg.com/originals/21/91/e2/2191e2da3e2236dcb0f2160bda37849c.gif

### OK, so what's in the consumer app, really?

It's pretty simple, and you don't need to be an expert in Go to read that code, so you can easily adapt it to your needs.

> Note: all the code snippets below are taken from [this repository](https://github.com/jotak/kafka-clickhouse-example/).

It reads messages from Kafka:

```
// ... running in a loop ...

		m, err := r.ReadMessage(context.Background())
		if err != nil {
			break
		}
		if flow, err := decode(m.Value); err != nil {
			klog.Errorf("failed to decode: %v", err)
		} else {
			out <- flow
		}

// ...

func decode(b []byte) (map[string]interface{}, error) {
	var rawLine map[string]interface{}
	err := json.Unmarshal(b, &rawLine)
	return rawLine, err
}
```

> It get fields from deserialized JSON:

Reviewer 2: ```suggestion
It gets fields from deserialized JSON:
```

```
	if v, ok := rawFlow[fields.Bytes]; ok {
		bytes = int(v.(float64))
	}
	if v, ok := rawFlow[fields.Packets]; ok {
		packets = int(v.(float64))
	}
	// etc.
```

And it executes the insert query:

```
	if err := conn.Exec(
		context.Background(),
		"INSERT INTO flows VALUES (?,?,?,?,?,?,?,?,?,?,?,?)",
		flowStart, flowEnd, srcAddr, dstAddr, srcName, dstName, srcKind, dstKind, srcNamespace, dstNamespace, bytes, packets,
	); err != nil {
		klog.Warnf("Insertion error: %v", err)
	}
```

That's about it.

It only maps a subset of the available flow data. You can check the [JSON reference](https://docs.openshift.com/container-platform/latest/networking/network_observability/json-flows-format-reference.html) to learn about all the fields.

> If you want to adapt it in a secure environment, you will need to configure the [kafka-go client](https://github.com/segmentio/kafka-go#tls-support) for TLS/mTLS/SASL, and the [clickhouse-go client](https://github.com/ClickHouse/clickhouse-go#tlsssl) for TLS and setting up credentials.

Reviewer 2: ```suggestion
If you want to adapt it in a secure environment, you need to configure the [kafka-go client](https://github.com/segmentio/kafka-go#tls-support) for TLS/mTLS/SASL, and the [clickhouse-go client](https://github.com/ClickHouse/clickhouse-go#tlsssl) for TLS and setting up credentials.
```
